pkg/controllers/state/metrics/node.go

```diff
@@ -0,0 +1,246 @@
+		}
+
+		return true
+	})
+
+	ns.cleanup(ctx, nodes)
+}
+
+func (ns *nodeScraper) reset() {
+	for _, gauge := range []*prometheus.GaugeVec{
+		allocatableGaugeVec,
+		podRequestsGaugeVec,
+		podLimitsGaugeVec,
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		gauge.Reset()
+	}
+}
+
+func (ns *nodeScraper) cleanup(ctx context.Context, existingNodes map[string]struct{}) {
+	nodesToRemove := []string{}
+	for nodeName := range ns.labelMap {
+		if _, ok := existingNodes[nodeName]; !ok {
+			nodesToRemove = append(nodesToRemove, nodeName)
+		}
+	}
+
+	// Remove all gauges associated with removed node
+	for _, node := range nodesToRemove {
+		gaugeMap := ns.labelMap[node]
+		for gaugeVec, labelSet := range gaugeMap {
+			for _, labels := range labelSet {
+				gaugeVec.Delete(labels)
+			}
+		}
+		delete(ns.labelMap, node)
+	}
+
+	if len(nodesToRemove) > 0 {
+		logging.FromContext(ctx).Debugf("Removed the following node gauges: %s", strings.Join(nodesToRemove, ", "))
+	}
+}
+
+// set sets the value for the node gauge
+func (ns *nodeScraper) set(resourceList v1.ResourceList, node *v1.Node, gaugeVec *prometheus.GaugeVec) error {
+	for resourceName, quantity := range resourceList {
+		resourceTypeName := strings.ReplaceAll(strings.ToLower(string(resourceName)), "-", "_")
+		labels := ns.getNodeLabels(node, resourceTypeName)
+		ns.labelMap[node.Name][gaugeVec] = append(ns.labelMap[node.Name][gaugeVec], labels)
+
+		gauge, err := gaugeVec.GetMetricWith(labels)
+		if err != nil {
+			return fmt.Errorf("generating new gauge: %w", err)
+		}
+		if resourceName == v1.ResourceCPU {
+			gauge.Set(float64(quantity.MilliValue()) / float64(1000))
+		} else {
+			gauge.Set(float64(quantity.Value()))
+		}
+	}
+	return nil
+}
+
+func (ns *nodeScraper) getSystemOverhead(node *v1.Node) v1.ResourceList {
+	systemOverheads := v1.ResourceList{}
+	if len(node.Status.Allocatable) > 0 {
+		// calculating system daemons overhead
+		for resourceName, quantity := range node.Status.Allocatable {
+			overhead := node.Status.Capacity[resourceName]
+			overhead.Sub(quantity)
+			systemOverheads[resourceName] = overhead
+		}
+	}
+	return systemOverheads
+}
+
+func (ns *nodeScraper) getNodeLabels(node *v1.Node, resourceTypeName string) prometheus.Labels {
+	metricLabels := prometheus.Labels{}
+	metricLabels[resourceType] = resourceTypeName
+	metricLabels[nodeName] = node.GetName()
+	if provisionerName, ok := node.Labels[v1alpha5.ProvisionerNameLabelKey]; !ok {
+		metricLabels[nodeProvisioner] = "N/A"
+	} else {
+		metricLabels[nodeProvisioner] = provisionerName
+	}
+	metricLabels[nodePhase] = string(node.Status.Phase)
+
+	// Populate well known labels
+	for wellKnownLabel, label := range getWellKnownLabels() {
+		if value, ok := node.Labels[wellKnownLabel]; !ok {
+			metricLabels[label] = "N/A"
+		} else {
+			metricLabels[label] = value
+		}
+	}
+
+	return metricLabels
+}
+
+func getWellKnownLabels() map[string]string {
```

Comment:
Is it possible to generate this statically so that we don't need to recompute this on line 225?

`var WellKnownLabels = getWellKnownLabels()`

and then use `WellKnownLabels` everywhere.